### PR TITLE
Update to Rust 1.78

### LINKED ISSUE #32413: mach doc hangs and eventually crashes on wgpu-core(doc)
The hang also happens locally, not sure about the crash.

- Last good build: <https://github.com/servo/servo/actions/runs/8837709219>
- First bad build: <https://github.com/servo/servo/actions/runs/8845098367>

### PR DESCRIPTION
**Pull request by sagudev**
Somebody asked for it.

Everybody will need to rerun `mach bootstrap` to rebuild crown (can we do this automatically?).

Requires new version of ipc-channel.

Fixes #32413 

---

- [x] `./mach build -d` does not report any errors
- [x] `./mach test-tidy` does not report any errors
- [x] There are tests for these changes (compiler and crown tests)

### CODE REVIEW

**Comment by sagudev** [comment]
The only problem that I have now is updating this in nix, we need to wait for https://github.com/oxalica/rust-overlay/tree/master to get new version.

File: support/crown/src/unrooted_must_root.rs [hunk 1]
```diff
@@ -331,7 +336,9 @@ impl<'a, 'tcx> visit::Visitor<'tcx> for FnDefVisitor<'a, 'tcx> {
                     cx.lint(
                         UNROOTED_MUST_ROOT,
                         format!("Expression of type {:?} must be rooted", ty),
-                        |lint| lint.set_span(pat.span),
+                        |lint| {
```
Comment (mrobinson) [inline support/crown/src/unrooted_must_root.rs:339]: This change looks unrelated?

File: support/crown/src/common.rs [hunk 2]
```diff
@@ -238,6 +247,19 @@ pub fn def_path_res(cx: &LateContext<'_>, path: &[&str]) -> Vec<Res> {
...
+        .find_map(|res| match res {
+            Res::Def(DefKind::Trait | DefKind::TraitAlias, trait_id) => Some(trait_id),
+            _ => None,
+        })
+}
+
+/*
+These are special variants made from above functions
+*/
+
```
Comment (mrobinson) [inline support/crown/src/common.rs:262]:
```suggestion
// These are special variants made from the above functions.
```

File: support/crown/Cargo.toml [hunk 3]
```diff
@@ -20,3 +20,7 @@ trace_in_no_trace_lint = []
 [package.metadata.rust-analyzer]
 # This crate uses #![feature(rustc_private)]
 rustc_private = true
+
+# You might also need this in .vscode/settings.json
+# "rust-analyzer.check.command": "check",
+# "rust-analyzer.rustc.source": "discover",
```
Comment (mrobinson) [inline support/crown/Cargo.toml:26]: Can you please also update https://github.com/servo/servo/blob/main/docs/HACKING_QUICKSTART.md#visual-studio-code if necessary?

**Review by mrobinson — APPROVED** [review]

File: support/crown/src/unrooted_must_root.rs (hunk 1 above)
Comment (sagudev) [reply support/crown/src/unrooted_must_root.rs:339]: `.with_span` (or what was before `.set_span`) started to return `&mut Self` while closure expects `()` for return type.

File: support/crown/Cargo.toml (hunk 3 above)
Comment (sagudev) [reply support/crown/Cargo.toml:26]: This slows down rust-analyzer (that's why it isn't set by default) and it's is only really needed when working on crown, so it's mostly not needed (even I do not have it when not working on crown).

**Comment by sagudev** [comment]
@mrobinson land it, if it is OK.

File: support/crown/Cargo.toml [hunk 4]
```diff
@@ -20,3 +20,7 @@ trace_in_no_trace_lint = []
 [package.metadata.rust-analyzer]
 # This crate uses #![feature(rustc_private)]
 rustc_private = true
+
+# You might also need this in .vscode/settings.json
```
Comment (mrobinson) [inline support/crown/Cargo.toml:24]:
```suggestion
# If you are working on crown, you might also need this in .vscode/settings.json.
```

File: support/crown/src/common.rs [hunk 5]
```diff
@@ -69,8 +70,8 @@ macro_rules! symbols {
 Stuff copied from clippy:
 */
 
+// https://github.com/rust-lang/rust-clippy/blob/546408be416f0355a39601c1457b37727bc74395/clippy_utils/src/lib.rs#L517
```
Comment (mrobinson) [inline support/crown/src/common.rs:73]:
```suggestion
// This is adapted from
// https://github.com/rust-lang/rust-clippy/blob/546408be416f0355a39601c1457b37727bc74395/clippy_utils/src/lib.rs#L517.
```

File: support/crown/src/common.rs [hunk 6]
```diff
@@ -121,16 +128,17 @@ fn non_local_item_children_by_name(tcx: TyCtxt<'_>, def_id: DefId, name: Symbol)
     }
 }
 
+// from clippy: https://github.com/rust-lang/rust-clippy/blob/546408be416f0355a39601c1457b37727bc74395/clippy_utils/src/lib.rs#L574
```
Comment (mrobinson) [inline support/crown/src/common.rs:131]:
```suggestion
// This is adapted from clippy:
// https://github.com/rust-lang/rust-clippy/blob/546408be416f0355a39601c1457b37727bc74395/clippy_utils/src/lib.rs#L574.
```

**Review by mrobinson — APPROVED** [review]

**Comment by sagudev** [comment]
CRASH [expected OK] /_webgl/conformance/more/functions/texImage2DBadArgs.html
    CRASH [expected OK] /_webgl/conformance/textures/misc/gl-teximage.html

This could be due to LLVM bump in rust.

**Comment by jschwe** [comment]
```
   │ unsafe precondition(s) violated: ptr::copy_nonoverlapping requires that both pointer arguments are aligned and non-null and the specified memory ranges do not overlap (thread Script(1,1), at library/core/src/panicking.rs:156)
```

I believe its the debug assertions being enabled in the standard library now in Rust 1.78.

**Comment by sagudev** [comment]
`_webgl/conformance/more/functions/texImage2DBadArgs.html`:
```
unsafe precondition(s) violated: ptr::copy_nonoverlapping requires that both pointer arguments are aligned and non-null and the specified memory ranges do not overlap (thread Script(1,1), at library/core/src/panicking.rs:156)
 0:02.48 pid:62526    0: servoshell::backtrace::print
 0:02.48 pid:62526    1: servoshell::main::{{closure}}
 0:02.48 pid:62526    2: std::panicking::rust_panic_with_hook
 0:02.48 pid:62526    3: std::panicking::begin_panic_handler::{{closure}}
 0:02.48 pid:62526    4: std::sys_common::backtrace::__rust_end_short_backtrace
 0:02.48 pid:62526    5: rust_begin_unwind
 0:02.48 pid:62526    6: core::panicking::panic_nounwind_fmt
 0:02.48 pid:62526    7: core::panicking::panic_nounwind
 0:02.48 pid:62526    8: ipc_channel::platform::unix::OsIpcSharedMemory::from_bytes
 0:02.48 pid:62526    9: ipc_channel::ipc::IpcSharedMemory::from_bytes
 0:02.48 pid:62526   10: <script::dom::webglrenderingcontext::WebGLRenderingContext as script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContext_Binding::WebGLRenderingContextMethods>::TexImage2D
 0:02.48 pid:62526   11: script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContext_Binding::texImage2D::{{closure}}
 0:02.48 pid:62526   12: mozjs::panic::wrap_panic
 0:02.48 pid:62526   13: script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContext_Binding::texImage2D
 0:02.48 pid:62526   14: CallJitMethodOp
 0:02.48 pid:62526   15: script::dom::bindings::utils::generic_call
 0:02.48 pid:62526   16: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:02.48 pid:62526   17: _ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
 0:02.48 pid:62526   18: _ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
 0:02.48 pid:62526   19: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:02.48 pid:62526   20: _ZN2js9InterpretEP9JSContextRNS_8RunStateE
 0:02.48 pid:62526   21: _ZN2js9RunScriptEP9JSContextRNS_8RunStateE
 0:02.48 pid:62526   22: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:02.48 pid:62526   23: _ZN2js4CallEP9JSContextN2JS6HandleINS2_5ValueEEES5_RKNS_13AnyInvokeArgsENS2_13MutableHandleIS4_EENS_10CallReasonE
 0:02.48 pid:62526   24: _ZN2js9fun_applyEP9JSContextjPN2JS5ValueE
 0:02.48 pid:62526   25: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:02.48 pid:62526   26: _ZN2js3jit14DoCallFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_14ICFallbackStubEjPN2JS5ValueENS7_13MutableHandleIS8_EE
 0:02.48 pid:62526   27: <unknown>
 0:02.48 pid:62526   28: <unknown>
 0:02.48 pid:62526   29: <unknown>
 0:02.48 pid:62526   30: <unknown>
 0:02.48 pid:62526   31: _ZN2js3jit32EnterBaselineInterpreterAtBranchEP9JSContextPNS_16InterpreterFrameEPh
 0:02.48 pid:62526   32: <unknown>
 0:02.48 pid:62526 mozilla::detail::MutexImpl::~MutexImpl: pthread_mutex_destroy failed: Device or resource busy
 0:02.48 pid:62526 Caught signal 11thread panicked while processing panic. aborting.
 0:02.49 pid:62526 Redirecting call to abort() to mozalloc_abort
 0:02.49 pid:62526 
 0:02.61 TEST_END: CRASH, expected OK
```
`_webgl/conformance/textures/misc/gl-teximage.html`
```
pid:60975 unsafe precondition(s) violated: ptr::copy_nonoverlapping requires that both pointer arguments are aligned and non-null and the specified memory ranges do not overlap (thread Script(1,1), at library/core/src/panicking.rs:156)
 0:03.09 pid:60975    0: servoshell::backtrace::print
 0:03.09 pid:60975    1: servoshell::main::{{closure}}
 0:03.09 pid:60975    2: std::panicking::rust_panic_with_hook
 0:03.09 pid:60975    3: std::panicking::begin_panic_handler::{{closure}}
 0:03.09 pid:60975    4: std::sys_common::backtrace::__rust_end_short_backtrace
 0:03.09 pid:60975    5: rust_begin_unwind
 0:03.09 pid:60975    6: core::panicking::panic_nounwind_fmt
 0:03.09 pid:60975    7: core::panicking::panic_nounwind
 0:03.09 pid:60975    8: ipc_channel::platform::unix::OsIpcSharedMemory::from_bytes
 0:03.10 pid:60975    9: ipc_channel::ipc::IpcSharedMemory::from_bytes
 0:03.10 pid:60975   10: <script::dom::webglrenderingcontext::WebGLRenderingContext as script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContext_Binding::WebGLRenderingContextMethods>::TexImage2D
 0:03.10 pid:60975   11: script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContext_Binding::texImage2D::{{closure}}
 0:03.10 pid:60975   12: mozjs::panic::wrap_panic
 0:03.10 pid:60975   13: script::dom::bindings::codegen::Bindings::WebGLRenderingContextBinding::WebGLRenderingContext_Binding::texImage2D
 0:03.10 pid:60975   14: CallJitMethodOp
 0:03.10 pid:60975   15: script::dom::bindings::utils::generic_call
 0:03.10 pid:60975   16: _ZN2js23InternalCallOrConstructEP9JSContextRKN2JS8CallArgsENS_14MaybeConstructENS_10CallReasonE
 0:03.10 pid:60975   17: _ZN2js3jit14DoCallFallbackEP9JSContextPNS0_13BaselineFrameEPNS0_14ICFallbackStubEjPN2JS5ValueENS7_13MutableHandleIS8_EE
 0:03.10 pid:60975   18: <unknown>
 0:03.10 pid:60975   19: <unknown>
 0:03.10 pid:60975   20: <unknown>
 0:03.10 pid:60975   21: <unknown>
 0:03.10 pid:60975   22: _ZN2js3jit32EnterBaselineInterpreterAtBranchEP9JSContextPNS_16InterpreterFrameEPh
 0:03.10 pid:60975 mozilla::detail::MutexImpl::~MutexImpl: pthread_mutex_destroy failed: Device or resource busy
 0:03.10 pid:60975 Caught signal 11thread panicked while processing panic. aborting.
 0:03.10 pid:60975 Redirecting call to abort() to mozalloc_abort
 0:03.10 pid:60975 
 0:03.24 TEST_END: CRASH, expected OK
```

**Comment by sagudev** [comment]
> ```
>    │ unsafe precondition(s) violated: ptr::copy_nonoverlapping requires that both pointer arguments are aligned and non-null and the specified memory ranges do not overlap (thread Script(1,1), at library/core/src/panicking.rs:156)
> ```
> 
> I believe its the debug assertions being enabled in the standard library now in Rust 1.78.

Right, that makes much more sense.

**Comment by sagudev** [comment]
Apparently destination pointer is null although it shouldn't be as this was checked upon it's creation: https://github.com/sagudev/ipc-channel/blob/4c01b78a108d761ac67407da1359b8293c7c93ef/src/platform/unix/mod.rs#L803

**Comment by sagudev** [comment]
> Note that even if the effectively copied size (count * size_of::<T>()) is 0, the pointers must be non-null and properly aligned.

**Comment by sagudev** [comment]
Related: https://github.com/servo/ipc-channel/issues/268

**Comment by sagudev** [comment]
Testing https://github.com/servo/ipc-channel/pull/335

**Comment by sagudev** [comment]
I guess there are even more debug assert failures, this time unrelated to ipc-channel. (I somehow knew, that it was to good to be true for whole repo to be good; although why they were not triggered before? rebase?).

**Comment by sagudev** [comment]
TODO: Do fixme from https://github.com/servo/servo/pull/32262.

**Comment by sagudev** [comment]
> @sagudev Is [servo/ipc-channel#335](https://github.com/servo/ipc-channel/pull/335) the last thing that is blocking the Rust 1.78 upgrade, or are there any other failures?

https://github.com/servo/ipc-channel/pull/335 is the only blocker (we also need to release new version for it).

**Comment by sagudev** [comment]
Rebased.

**Review by mrobinson — APPROVED** [review]
